### PR TITLE
Add Evmos to Dockernet

### PR DESCRIPTION
**Pull request by asalzmann**
**Summary**
Adds Evmos as a dockernet host zone. Evmos gives us coverage around additioonal edge cases that are not caught from standard cosmos chains:
* Ethermint based
* Non-118 coin type
* Different micro denom granuarity

Given Evmos is ethermint based, a few additional changes were required in this PR
* increase tokens allocated to validators at genesis (1evmos = 1e18evmos)
* Update withdrawal callback to handle new balance type (sdk 46 uses Ints, not Coins to store balances) (handled in an earlier PR)

**Test plan**

run integration tests
```
make start-docker
make test-integration-docker
```

result
```
integration_tests.bats
 ✓ [INTEGRATION-BASIC-EVMOS] host zones successfully registered
 ✓ [INTEGRATION-BASIC-EVMOS] ibc transfer updates all balances
 ✓ [INTEGRATION-BASIC-EVMOS] liquid stake mints stToken
 ✓ [INTEGRATION-BASIC-EVMOS] tokens were transferred to evmos_9001-2 after liquid staking
 ✓ [INTEGRATION-BASIC-EVMOS] tokens on EVMOS were staked
 ✓ [INTEGRATION-BASIC-EVMOS] redemption works
 ✓ [INTEGRATION-BASIC-EVMOS] claimed tokens are properly distributed
 ✓ [INTEGRATION-BASIC-EVMOS] rewards are being reinvested, exchange rate updating
 ✓ [INTEGRATION-BASIC-EVMOS] revenue accrued, and clear-balance works

9 tests, 0 failures
```

simulate acks
ica transfer - timeout
```
# force timeout in `TransferExistingDepositsToHostZones`
make start-docker && make test-integration-docker

# expect: deposit record state reverts to TRANSFER_QUEUE
# observed: 
deposit_record:
- amount: "100000"
  denom: aevmos
  deposit_epoch_number: "3"
  host_zone_id: evmos_9001-2
  id: "1"
  source: STRIDE
  status: TRANSFER_QUEUE
```

ica delegate - error
```
# set invalid (unregistered) validator address in `DelegateOnHost`

# expect: deposit record state reverts to DELEGATION_QUEUE
# observed: ack failure and deposit record status reversion 
dockernet-stride1-1  | 8:38PM ERR |   EVMOS_9001-2  |  DELEGATE ICACALLBACK  |  ICA FAILED (ack error), Packet: {Sequence:6 SourcePort:icacontroller-evmos_9001-2.DELEGATION SourceChannel:channel-1 DestinationPort:icahost ...
deposit_record:
- amount: "100000"
  denom: aevmos
  deposit_epoch_number: "3"
  host_zone_id: evmos_9001-2
  id: "1"
  source: STRIDE
  status: DELEGATION_QUEUE

```

ica delegate - timeout
```
# force timeout in `SubmitTxs`
# expect: deposit record state reverts to DELEGATION_QUEUE
# observed: 
dockernet-stride1-1  | 10:07PM ERR |   EVMOS_9001-2  |  DELEGATE ICACALLBACK  |  ICA TIMEOUT, Packet: {Sequence:4 SourcePort:icacontroller-evmos_9001-2.DELEGATION SourceChannel:channel-1 DestinationPort:icahost ...

deposit_record:
- amount: "100000"
  denom: aevmos
  deposit_epoch_number: "3"
  host_zone_id: evmos_9001-2
  id: "1"
  source: STRIDE
  status: DELEGATION_IN_PROGRESS

# this status is expected and reverts by running `RestoreInterchainAccount` 
```

### CODE REVIEW

File: dockernet/config/evmos_ica.json [hunk 1]
```diff
@@ -0,0 +1,24 @@
+{
+    "interchainaccounts": {
+        "host_genesis_state": {
+            "active_channels": [],
+            "interchain_accounts": [],
+            "port": "icahost",
+            "params": {
+                "host_enabled": true,
+                "allow_messages": [
```
Comment (fedekunze) [inline dockernet/config/evmos_ica.json:9]: this seems to be a different set of messages compared to the ones we'll support

File: dockernet/config/evmos_ica.json [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+                    "/cosmos.bank.v1beta1.MsgSend",
+                    "/cosmos.bank.v1beta1.MsgMultiSend",
+                    "/cosmos.staking.v1beta1.MsgDelegate",
+                    "/cosmos.staking.v1beta1.MsgUndelegate",
+                    "/cosmos.staking.v1beta1.MsgBeginRedelegate",
+                    "/cosmos.staking.v1beta1.MsgRedeemTokensforShares",
+                    "/cosmos.staking.v1beta1.MsgTokenizeShares",
```
Comment (fedekunze) [inline dockernet/config/evmos_ica.json:16]: not supported atm
```suggestion
```

File: dockernet/config/relayer_config.yaml [hunk 3]
```diff
@@ -74,6 +74,22 @@ chains:
...
+  evmos:
+    type: cosmos
+    value:
+      key: rly6
+      chain-id: evmos_9001-2
+      rpc-addr: http://evmos1:26657
+      account-prefix: evmos
+      keyring-backend: test
+      gas-adjustment: 1.2
+      gas-prices: 0.01uevmos
```
Comment (fedekunze) [inline dockernet/config/relayer_config.yaml:86]: denom is wrong here, should be "aevmos"

File: dockernet/src/app.toml [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+###############################################################################
+###                           Base Configuration                            ###
+###############################################################################
+
+# COPIED
+
+# The minimum gas prices a validator is willing to accept for processing a
+# transaction. A transaction's fees must meet the minimum of any denomination
+# specified in this config (e.g. 0.25token1;0.0001token2).
+minimum-gas-prices = "0uevmos"
```
Comment (fedekunze) [inline dockernet/src/app.toml:13]:
```suggestion
minimum-gas-prices = "20000000000aevmos"
```

**Review by fedekunze — COMMENTED** [review]
The Evmos config around host params and denomination is incorrect. See comments

File: dockernet/config/evmos_ica.json (hunk 2 above)
Comment (asalzmann) [reply dockernet/config/evmos_ica.json:16]: thanks, updated to reflect https://github.com/evmos/evmos/blob/main/app/upgrades/v11/upgrades.go#L77

File: dockernet/config/evmos_ica.json (hunk 1 above)
Comment (asalzmann) [reply dockernet/config/evmos_ica.json:9]: updated

File: dockernet/config/relayer_config.yaml (hunk 3 above)
Comment (asalzmann) [reply dockernet/config/relayer_config.yaml:86]: updated

File: dockernet/config.sh [hunk 5]
```diff
@@ -40,6 +40,8 @@ STATOM_DENOM="stuatom"
 STJUNO_DENOM="stujuno"
 STOSMO_DENOM="stuosmo"
 STSTARS_DENOM="stustars"
+EVMOS_DENOM="uevmos"
+STEVMOS_DENOM="stuevmos"
```
Comment (MalteHerrmann) [inline dockernet/config.sh:44]:
```suggestion
EVMOS_DENOM="aevmos"
STEVMOS_DENOM="staevmos"
```

**Review by MalteHerrmann — COMMENTED** [review]
Found one more instance of the `uevmos` denom floating around

**Comment by asalzmann** [comment]
Yep, will re-run everything on `ec337e0357ae6b6727ad9092ac0ce3667df91319` @danburck

**Comment by asalzmann** [comment]
Unfortunately there are some bugs with the go relayer

#1 error sending messages
```
dockernet-relayer-evmos-1  | ts=2023-01-13T04:11:22.591446Z lvl=error msg="Error sending messages" src_chain_id=evmos_9001-2 dst_chain_id=STRIDE src_client_id=07-tendermint-0 dst_client_id=07-tendermint-0 messagesError="unsupported value type" error="unable to resolve type URL /stride.interchainquery.v1.MsgSubmitQueryResponse: tx parse error" errorVerbose="\ngithub.com/cosmos/cosmos-sdk/x/auth/tx.DefaultTxDecoder.func1\n\t/go/pkg/mod/github.com/cosmos/cosmos-sdk@v0.45.10/x/auth/tx/decoder.go:42\ngithub.com/strangelove-ventures/lens/client.mkTxResult\n\t/go/pkg/mod/github.com/!stride-!labs/lens@v0.5.2-patch-eth-cdc-1/client/broadcast.go:114\ngithub.com/strangelove-ventures/lens/client.broadcastTx\n\t/go/pkg/mod/github.com/!stride-!labs/lens@v0.5.2-patch-eth-cdc-1/client/broadcast.go:105\ngithub.com/strangelove-ventures/lens/client.(*ChainClient).BroadcastTx\n\t/go/pkg/mod/github.com/!stride-!labs/lens@v0.5.2-patch-eth-cdc-1/client/broadcast.go:36\ngithub.com/cosmos/relayer/v2/relayer/chains/cosmos.(*CosmosProvider).SendMessages.func1\n\t/src/relayer/relayer/chains/cosmos/tx.go:163\ngithub.com/avast/retry-go/v4.Do\n\t/go/pkg/mod/github.com/avast/retry-go/v4@v4.3.1/retry.go:128\ngithub.com/cosmos/relayer/v2/relayer/chains/cosmos.(*CosmosProvider).SendMessages\n\t/src/relayer/relayer/chains/cosmos/tx.go:87\ngithub.com/cosmos/relayer/v2/relayer/processor.(*PathProcessor).sendMessages\n\t/src/relayer/relayer/processor/path_processor_internal.go:923\nunable to resolve type URL /stride.interchainquery.v1.MsgSubmitQueryResponse: tx parse error"
```

#2 Error building or broadcasting transaction
```
dockernet-relayer-evmos-1  | ts=2023-01-13T04:11:30.508186Z lvl=info msg="Error building or broadcasting transaction" provider_type=cosmos chain_id=STRIDE attempt=1 max_attempts=5 error="rpc error: code = Unknown desc = rpc error: code = Unknown desc = failed to execute message; message index: 1: unable to unmarshal balance in callback args for zone: evmos_9001-2, err: proto: illegal wireType 7: unable to marshal data structure [Stride-Labs/stride/v4/x/stakeibc/keeper/icqcallbacks_withdrawal_balance.go:35] With gas wanted: '18446744073709551615' and gas used: '115782' : unknown request"
```

looking into those today

**Comment by asalzmann** [comment]
Resolved the issue (sdk 46 stores balances as `Int` types rather than `Coins`, so ICQ unmarshalling had to be updated)

**Comment by asalzmann** [comment]
Open

**Review by sampocs — APPROVED** [review]
lgtm!

**Pull request merged**